### PR TITLE
Handle operational errors

### PR DESCRIPTION
**Pull request by WinnyTroy**

### CODE REVIEW

File: onadata/libs/utils/middleware.py [hunk 1]
```diff
@@ -64,3 +75,20 @@ def __call__(self, request):
...
+    def process_exception(self, request, exception):
+        # Filter out OperationalError Exceptions
+        if isinstance(exception, OperationalError):
+            already_raised = getattr(settings, "ALREADY_RAISED", False)
+            if not already_raised:
+                settings.ALREADY_RAISED = True
+                with use_master:
+                    response = self.get_response(request)
+                    return response
+            settings.ALREADY_RAISED = None
```
Comment (lincmba) [inline onadata/libs/utils/middleware.py]: I'd advise keeping this value as a bool and hence have either `True` or`False` as its value and not `None`.

File: onadata/libs/utils/middleware.py [hunk 2]
```diff
@@ -64,3 +75,20 @@ def __call__(self, request):
...
+
+class OperationalErrorMiddleware(BaseMiddleware):
+    """
+    Captures requests returning 500 status code.
+    Then retry it against master database.
+    """
+    def process_exception(self, request, exception):
+        # Filter out OperationalError Exceptions
+        if isinstance(exception, OperationalError):
+            already_raised = getattr(settings, "ALREADY_RAISED", False)
```
Comment (lincmba) [inline onadata/libs/utils/middleware.py]: I believe keeping this value in the settings file means all subsequent errors after the first will be ignored as this is not tracking independent requests.

Comment (ukanga) [reply onadata/libs/utils/middleware.py]: I'm not sure if the setting traverses across multiple threads, an attempt to use `threading.local()` did not work. Using settings is what seemed to work.

File: onadata/libs/utils/middleware.py (hunk 1 above)
Comment (ukanga) [reply onadata/libs/utils/middleware.py]: I am okay with `False`.

File: onadata/libs/utils/middleware.py (hunk 2 above)
Comment (ukanga) [reply onadata/libs/utils/middleware.py]: I am not sure we should add the setting `ALREADY_RAISED` to the setting files, it should just have life in this middleware in the event it is used. Using `getattr()` with a default value is a valid use case with no required configuration.

Comment (lincmba) [reply onadata/libs/utils/middleware.py]: I agree. `ALREADY_RAISED` variable should be defined and only used in this middleware.

Comment (ukanga) [inline onadata/libs/utils/middleware.py]: I think we need to limit the middleware to very specific `OperationaError` exceptions. Specifically the ones that contain "canceling statement due to conflict with recovery".

**Pull request merged**